sound/soc/sof/intel/hda.c

```diff
@@ -53,67 +53,107 @@ static int sdw_clock_stop_quirks = SDW_INTEL_CLK_STOP_BUS_RESET;
 module_param(sdw_clock_stop_quirks, int, 0444);
 MODULE_PARM_DESC(sdw_clock_stop_quirks, "SOF SoundWire clock stop quirks");
 
+int hda_dai_widget_update(struct snd_sof_dev *sdev, struct snd_sof_dai *sof_dai,
+			  struct sof_ipc_dai_config *config, bool setup, int dir)
+{
+	struct snd_sof_widget *swidget;
+	struct sof_ipc_reply reply;
+	bool found = false;
+	int ret;
+
+	/* find snd_sof_widget of type DAI with matching sname as that of snd_sof_dai name */
+	list_for_each_entry(swidget, &sdev->widget_list, list) {
+		if (dir == SNDRV_PCM_STREAM_PLAYBACK && swidget->id != snd_soc_dapm_dai_in)
+			continue;
+
+		if (dir == SNDRV_PCM_STREAM_CAPTURE && swidget->id != snd_soc_dapm_dai_out)
+			continue;
+
+		if (!strcmp(sof_dai->name, swidget->widget->sname)) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) {
+		dev_err(sdev->dev, "error: failed to find DAI widget for: %s\n",
+			sof_dai->cpu_dai_name);
+		return -EINVAL;
+	}
+
+	/*
+	 * For static pipelines, the DAI widget would already be set up and calling
+	 * sof_widget_setup()/free() would simply return without doing anything.
+	 * For dynamic pipelines, the DAI widget would be set up or freed here.
+	 */
+	if (setup)
+		ret = sof_widget_setup(sdev, swidget);
+	else
+		ret = sof_widget_free(sdev, swidget);
+	if (ret < 0)
+		return ret;
+
+	/* send DAI_CONFIG IPC */
+	return sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size, &reply,
+				  sizeof(reply));
+}
+
+static int sdw_dai_config_ipc(struct snd_sof_dev *sdev, struct snd_soc_dai *d,
+			      int link_id, int alh_stream_id, bool setup, int dir)
+{
+	struct sof_ipc_dai_config *config;
+	struct snd_sof_dai *sof_dai;
+	bool found = false;
+
+	/* find snd_sof_dai matching the BE CPU DAI name */
+	list_for_each_entry(sof_dai, &sdev->dai_list, list) {
+		if (!sof_dai->cpu_dai_name)
+			continue;
+
+		if (!strcmp(d->name, sof_dai->cpu_dai_name)) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) {
+		dev_err(sdev->dev, "error: failed to find SOF DAI for: %s\n", d->name);
+		return -EINVAL;
+	}
+
+	config = sof_dai->dai_config;
+	if (!config) {
+		dev_err(sdev->dev, "error: no config for DAI %s\n", sof_dai->name);
+		return -EINVAL;
+	}
+
+	/* update config with link and stream ID */
+	config->dai_index = (link_id << 8) | d->id;
+	config->alh.stream_id = alh_stream_id;
+
+	return hda_dai_widget_update(sdev, sof_dai, config, setup, dir);
```

Comment:
the naming doesn't seem too good, why would SoundWire have anything to do with hda? the prefix looks odd and wrong at the same time.